src/smc-webapp/project/new/project-new-form.tsx

```diff
@@ -0,0 +1,433 @@
+
+import { React, rtypes, rclass } from "../../app-framework";
+import { default_filename } from "../../account";
+
+import {
+  Col,
+  Row,
+  Button,
+  ButtonToolbar,
+  FormControl,
+  FormGroup,
+  Alert
+} from "react-bootstrap";
+
+import { ErrorDisplay, Icon, Tip, SettingBox } from "../../r_misc";
+import { ProjectActions } from "../../project_actions";
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const { special_filenames_with_no_extension } = require("../../project_file");
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const { SMC_Dropzone } = require("../../smc-dropzone");
+import { PathLink } from "./path-link";
+import { NewFileButton } from "./new-file-button";
+import { NewFileDropdown } from "./new-file-dropdown";
+import { FileTypeSelector } from "./file-type-selector";
+import { AvailableFeatures } from "./types";
+import { ProjectMap } from "smc-webapp/todo-types";
+
+interface ReactProps {
+  project_id: string;
+  actions: ProjectActions;
+  on_close?: () => void;
+  on_create_file?: () => void;
+  on_create_folder?: () => void;
+  show_header?: boolean;
+  default_filename?: string;
+  name: string;
+}
+
+interface ReduxProps {
+  current_path?: string;
+  default_filename: string;
+  file_creation_error: string;
+  available_features: AvailableFeatures;
+  project_map: ProjectMap;
+  downloading_file: boolean;
+  get_total_project_quotas: (project_id: string) => Record<string, any>;
+}
+
+interface State {
+  filename: string;
+  extension_warning: boolean;
+}
+
+export const ProjectNewForm = rclass(
+  class ProjectNewForm extends React.Component<ReactProps & ReduxProps, State> {
+    static reduxProps({ name }): any {
+      return {
+        [name]: {
+          current_path: rtypes.string,
+          default_filename: rtypes.string,
+          file_creation_error: rtypes.string,
+          available_features: rtypes.immutable,
+          downloading_file: rtypes.boolean
+        },
+        projects: {
+          project_map: rtypes.immutable,
+          get_total_project_quotas: rtypes.func
+        }
+      };
+    }
+
+    constructor(props) {
+      super(props);
+      this.state = {
+        filename: this.props.default_filename ?? this.default_filename(),
+        extension_warning: false
+      };
+    }
+
+    static defaultProps = { show_header: true };
+
+    UNSAFE_componentWillReceiveProps(newProps): void {
+      if (newProps.default_filename !== this.props.default_filename) {
+        this.setState({ filename: newProps.default_filename });
+      }
+    }
+
+    default_filename = (): string => {
+      return default_filename(undefined, this.props.project_id);
+    };
+
+    create_file = (ext?: string): void => {
+      if (!this.state.filename) {
+        return;
+      }
+      this.props.actions.create_file({
+        name: this.state.filename,
+        ext,
+        current_path: this.props.current_path
+      });
+      this.props.on_create_file?.();
```

Comment:
First use in TS of [optional chaining](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#optional-chaining). Since @williamstein has reservations about optional chaining, I looked for some [arguments against](https://medium.com/@vcarl/overly-defensive-programming-e7a1b3d234c2) optional chaining. 

While our Coffeescript code definitely exhibited this bad pattern. I think having to write types now makes the feature usable.

Reply:
I guess another not insignificant in favor of optional chaining is that it will make rewriting our existing coffeescript code easier...